plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,284 @@
+      Apply _ fun arg       -> computeL (FrameAppArg (Closure arg env) : ctx)     heap (Closure fun env)
+      Wrap ann tyn ty term' -> computeL (FrameWrap ann tyn ty : ctx)              heap (Closure term' env)
+      Unwrap _ term'        -> computeL (FrameUnwrap : ctx)                       heap (Closure term' env)
+      TyAbs{}               -> returnL  ctx heap cl
+      LamAbs{}              -> returnL  ctx heap cl
+      Builtin{}             -> returnL  ctx heap cl
+      Constant{}            -> returnL  ctx heap cl
+      Error{}               -> Failure
+      Var _ name            -> let l = lookupHeapLoc name env
+                               in case lookupHeap l heap of
+                                    Evaluated cl'   -> returnL ctx heap cl'
+                                    Unevaluated cl' -> computeL (FrameHeapUpdate l : ctx) heap cl'
+
+
+-- | Return a closure containing a value. Ideally the fact that we've
+-- got a value would be enforced in the (Haskell) type. Values still
+-- have to be contained in closures.  We could have something like
+-- @v= \x.y@, where @y@ is bound in an environment.  If we ever go on
+-- to apply @v@, we'll need the value of @y@.
+returnL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+returnL [] heap res                                                        = Success res heap
+returnL (FrameTyInstArg ty        : ctx) heap cl                           = instantiateEvaluate ctx heap ty cl
+returnL (FrameAppArg argClosure   : ctx) heap cl                           = evaluateFun ctx heap cl argClosure
+returnL (FrameWrap ann tyn ty     : ctx) heap (Closure v env)              = returnL ctx heap $ Closure (Wrap ann tyn ty v) env
+returnL (FrameHeapUpdate l        : ctx) heap cl                           = returnL ctx heap' cl
+                                                                                   where heap' = updateHeap l (Evaluated cl) heap
+                                                                                      -- Leave this out to make the machine call-by-name (really slow)
+returnL (FrameUnwrap              : ctx) heap (Closure (Wrap _ _ _ t) env) = returnL ctx heap (Closure t env)
+returnL (FrameUnwrap              : _)      _ (Closure term _)             = throw $ LMachineException "Attemtping to unwrap non-wrapped term" term
+
+
+-- | Apply a function to an argument and proceed.
+-- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+
+evaluateFun :: EvaluationContext -> Heap -> Closure -> Closure-> LMachineResult
```

Comment:
Passing closures around is indeed much nicer, I'd steal that to the CEK machine, but I'm not sure how this wrapping-unwrapping impacts performance.

Reply:
I found that using explicit closures made it a lot easier to keep track of what was going on, but you're right that it may slow things down a little (although the same is true of monads...).  I presume that ghc will eliminate the constructor and just use pairs internally, so it may not be too bad.   I could try to measure that some time and see if it makes any difference.

Reply:
Yes, it makes it a lot clearer. This machine's pretty inefficient for other reasons anyway, so getting rid of the closures wouldn't gain much here.  The question is whether there'd be any penalty if we used them in the CEK machine as well.

Reply:
Well, we can profile it. And probably nothing a few judicious `UNPACK` pragmas won't fix.